runtime/lib/registry/Registry.cpp

```diff
@@ -0,0 +1,50 @@
+#include <cstdint>
+#include <cstdio>
+#include <pybind11/pybind11.h>
+#include <unordered_map>
+
+#include "Types.h"
+
+namespace py = pybind11;
+
+// From PyBind11's documentation:
+//
+//     Do you have any global variables that are pybind11 objects or invoke pybind11 functions in
+//     either their constructor or destructor? You are generally not allowed to invoke any Python
+//     function in a global static context. We recommend using lazy initialization and then
+//     intentionally leaking at the end of the program.
+//
+// https://pybind11.readthedocs.io/en/stable/advanced/misc.html#common-sources-of-global-interpreter-lock-errors
+std::unordered_map<int64_t, py::function> *references;
+
+extern "C" {
+[[gnu::visibility("default")]] void callbackCall(int64_t identifier)
+{
+    auto it = references->find(identifier);
+    if (it == references->end()) {
+        throw std::invalid_argument("Callback called with invalid identifier");
+    }
+    auto lambda = it->second;
+    lambda();
+}
+}
+
+auto registerImpl(py::function f)
+{
+    // Do we need to see if it is already present or can we just override it?
+    // Does python reuse id's? Yes.
+    // But only after they have been garbaged collected.
+    // So as long as we maintain a reference to it, then they won't be garbage collected.
+    int64_t id = (int64_t)f.ptr();
+    references->insert({id, f});
+    return id;
+}
+
+PYBIND11_MODULE(registry, m)
+{
+    if (references == nullptr) {
+        references = new std::unordered_map<int64_t, py::function>();
```

Comment:
```suggestion
        references = std::make_unique<std::unordered_map<int64_t, py::function>>();
```

Reply:
This doesn't work as expected. Because now `references` is not a pointer but a struct (smart pointers are just structs), `references` will have a destructor that will be called at some point. This destructor normally gets called when the library is being closed. And for python modules, this likely happens after the process ends. E.g., the python interpreter is in a state where it does no longer have a main thread

```
Fatal Python error: PyThreadState_Get: the function must be called with the GIL held, but the GIL is released (the current Python thread state is NULL)
Python runtime state: finalizing (tstate=0x0000560ae7828740)
```

The destructor is attempting to call the destructor of the `py::function` objects, but there is no python runtime any more.

> [Destructs the unordered_map. The destructors of the elements are called and the used storage is deallocated. Note, that if the elements are pointers, the pointed-to objects are not destroyed. ](https://en.cppreference.com/w/cpp/container/unordered_map/%7Eunordered_map)